scripts/core/editor3/reducers/editor3.tsx

```diff
@@ -262,6 +269,46 @@ function escapeRegExp(string) {
     return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
 }
 
+function replaceText(
+    editorState: EditorState,
+    selection: SelectionState,
+    text: string,
+) {
+    const newContent = Modifier.replaceText(
+        editorState.getCurrentContent(),
+        selection,
+        text,
+    );
+
+    return EditorState.push(editorState, newContent, 'insert-characters');
+}
+
+function insertTab(state: EditorState, selection: SelectionState, block: ContentBlock, tabString: string) {
+    const indentInsteadOfInsert = !selection.isCollapsed();
+    const selectionForInsert = indentInsteadOfInsert
+        ? (selection.merge({
+            focusOffset: 0,
+            anchorOffset: 0,
+        }) as SelectionState)
+        : selection;
+
+    let newState = replaceText(state, selectionForInsert, tabString);
+
+    if (indentInsteadOfInsert) {
+        const restoreSelection = selectionForInsert.merge({
```

Comment:
Isn't it possible to replace text without losing the selection?

Reply:
It is, I explain it in the comment above